cmake: adding $<SEMICOLON> instead of space on zephyr_get_compile_<type>

tejlmand opened the pull request:
The current `zephyr_get_compile_<type>` flags returns compile / include
flags based on generator expressions.

But they also included a space.

This have the unintended side-effect that when the properties are
fetched into another variable and later appended to lists in order to
finally be used in a CMake target function / custom command, then CMake
in some cases will quote the string, creating a quoted argument like:
"arg0 arg1 arg2": instead of individual arguments like: "arg0" "arg1"
"arg2".

By using `$<SEMICOLON>`, then arguments are separated correctly.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: adding $<SEMICOLON> instead of space on zephyr_get_compile_<type>

The current zephyr_get_compile_<type> flags returns compile / include
flags based on generator expressions.

But they also included a space.

This have the unintended side-effect that when the properties are
fetched into another variable and later appended to lists in order to
finally be used in a CMake target function / custom command, then CMake
in some cases will quote the string, creating a quoted argument like:
"arg0 arg1 arg2: instead of individual arguments like: "arg0" "arg1"
"arg2".

By using $<SEMICOLON>, then arguments are seperated correctly.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: zephyr_get_<flags>_for_lang_as_string() uses space as delimiter

The use of zephyr_get_<flags>_for_lang_as_string will now use space as
delimiter, this will allow for existing used of the
zephyr_get_<flags>_for_lang_as_string() to return a string with spaces
instead of semicolons as a quoted list.

Note, in most situations it is prefered to have a proper CMake list
returned, in which case zephyr_get_<flags>_for_lang() is probably the
correct function to use.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

MaureenHelm commented:
Thanks @tejlmand ! This indeed fixes the issue in micropython/micropython#6542. Looks like buildkite crashed and burned though. I looked at one of the fails and was able to fix it like this:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9c13719792..9ed7c18733 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -754,7 +754,7 @@ if(CONFIG_CODE_DATA_RELOCATION)
 endif()
 
 if(CONFIG_USERSPACE)
-  zephyr_get_compile_options_for_lang(C compiler_flags_priv)
+  zephyr_get_compile_options_for_lang_as_string(C compiler_flags_priv)
   string(REPLACE "$<TARGET_PROPERTY:compiler,coverage>" ""
          NO_COVERAGE_FLAGS "${compiler_flags_priv}"
   )
```

cc: @dpgeorge

tejlmand committed:
cmake: using zephyr_get_compile_options_for_lang_as_string()

With the distinction between zephyr_get_compile_options_for_lang() and
zephyr_get_compile_options_for_lang_as_string() regarding list handling
and thus quoting, then zephyr_get_compile_options_for_lang_as_string()
is now used for CMake source file properties for compilation of
kobject_hash.c source file.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> Thanks @tejlmand ! This indeed fixes the issue in [micropython/micropython#6542](https://github.com/micropython/micropython/pull/6542). Looks like buildkite crashed and burned though. I looked at one of the fails and was able to fix it like this:

Great, thanks for confirming the fix.
I did notice the CI failure, but had to be away with family.

So thanks for the fix.
